database/src/config.rs

```diff
@@ -0,0 +1,200 @@
+    }
+}
+
+//---------------------------------------------------------------------------------------------------- SyncMode
+/// TODO
+#[derive(Copy, Clone, Default, PartialEq, PartialOrd, Eq, Ord, Hash)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+#[cfg_attr(
+    feature = "borsh",
+    derive(borsh::BorshSerialize, borsh::BorshDeserialize)
+)]
+pub enum SyncMode {
+    /// Fully sync to disk per transaction.
+    #[default]
+    Safe,
+
+    /// Asynchronously sync, only flush at database shutdown.
+    Fastest,
+}
+
+//---------------------------------------------------------------------------------------------------- ReaderThreads
+/// TODO
+#[derive(Copy, Clone, Default, PartialEq, PartialOrd)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+#[cfg_attr(
+    feature = "borsh",
+    derive(borsh::BorshSerialize, borsh::BorshDeserialize)
+)]
+pub enum ReaderThreads {
+    #[default]
+    /// TODO
+    OnePerThread,
+
+    /// TODO
+    One,
+
+    /// TODO
+    Number(NonZeroUsize),
+
+    /// TODO
+    ///
+    /// # Invariant
+    /// Must be `0.0..=1.0`.
+    Percent(f32),
+}
+
+impl ReaderThreads {
+    /// TODO
+    // # Invariant
+    // LMDB will error if we input zero, so don't allow that.
+    // <https://github.com/LMDB/lmdb/blob/b8e54b4c31378932b69f1298972de54a565185b1/libraries/liblmdb/mdb.c#L4687>
+    pub fn as_threads(&self) -> NonZeroUsize {
+        let total_threads = cuprate_helper::thread::threads();
+
+        match self {
+            Self::OnePerThread => total_threads,
+            Self::One => NonZeroUsize::MIN,
+            Self::Number(n) => std::cmp::min(*n, total_threads),
+
+            // We handle the casting loss.
+            #[allow(
+                clippy::cast_precision_loss,
+                clippy::cast_possible_truncation,
+                clippy::cast_sign_loss
+            )]
+            Self::Percent(f) => {
+                if !f.is_normal() || !(0.0..=1.0).contains(f) {
+                    return total_threads;
+                }
+
+                let thread_percent = (total_threads.get() as f32) * f;
+                let Some(threads) = NonZeroUsize::new(thread_percent as usize) else {
+                    return total_threads;
+                };
+
+                std::cmp::min(threads, total_threads)
+            }
+        }
+    }
+}
+
+/// TODO
+macro_rules! impl_from_reader_threads {
+    ($(
+        $from:ty // Type to convert into `ReaderThreads`
+    ),*) => {
+        $(
+            impl From<$from> for ReaderThreads {
+                fn from(value: $from) -> Self {
+                    match NonZeroUsize::new(value as usize) {
+                        Some(n) => Self::Number(n),
+                        None => Self::One,
+                    }
+                }
+            }
+        )*
+    };
+}
+impl_from_reader_threads!(u8, u16, u32, u64, usize, i8, i16, i32, i64, isize);
```

Comment:
you could do `TryInto` instead of `Into` and just panic

Reply:
`u16` is the highest unsigned int that can `Into<usize>`, so `ReaderThreads::from(1_u32)` won't work.

The reasoning behind casting is mostly for convenience since `ReaderThreads` must always be at least 1, the behavior for any input <1 is that it gets rounded up to 1.

Although, thinking about it the only input is going to be a `usize` from the user config? Going with `Into<usize>`.

Reply:
Actually how do feel about doing:
```rust
None => Self::OnePerThread
```
It would match `monerod` and most programs behavior of `0 == maximum value`.

Reply:
Yeah I think that would make sense, I am going to merge this as is, this can be added to a followup PR